src/address.rs

```diff
@@ -1,116 +1,90 @@
 use crate::error::Error;
 use crate::SubnetID;
+use fil_actors_runtime::cbor;
+use fvm_ipld_encoding::RawBytes;
 use fvm_shared::address::Address;
 use fvm_shared::ActorID;
 use lazy_static::lazy_static;
 use serde::{Deserialize, Serialize};
-use std::fmt;
-use std::fmt::Formatter;
 use std::str::FromStr;
 
-/// Maximum length for an address payload determined by
-/// the maximum size of the IPC address.
-const MAX_ADDRESS_LEN: usize = 54;
-
 // The default actor id namespace for IPC addresses
 lazy_static! {
     pub static ref DEFAULT_IPC_ACTOR_NAMESPACE_ID: ActorID = 1000u64;
 }
 
 #[derive(Clone, PartialEq, Eq, Debug, Hash, Serialize, Deserialize)]
 pub struct IPCAddress {
-    inner: Address,
-}
-
-impl fmt::Display for IPCAddress {
-    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
-        self.inner.fmt(f)
-    }
+    subnet_id: SubnetID,
+    raw_address: Address,
 }
 
 impl IPCAddress {
     /// Generates new address using ID protocol
-    pub const fn new_id(id: u64) -> Self {
+    pub fn new_id(id: u64) -> Self {
+        let d = SubnetID::default();
         Self {
-            inner: Address::new_id(id),
+            subnet_id: d,
+            raw_address: Address::new_id(id),
         }
     }
 
     /// Generates new IPC address
     pub fn new(sn: &SubnetID, addr: &Address) -> Result<Self, Error> {
-        let sn = sn.to_bytes();
-        let addr = addr.to_bytes();
-        let sn_size_vec = to_leb_bytes(sn.len() as u64)?;
-        let sn_size: &[u8] = sn_size_vec.as_ref();
-        let addr_size_vec = to_leb_bytes(addr.len() as u64)?;
-        let addr_size: &[u8] = addr_size_vec.as_ref();
-        let sp = [sn_size, addr_size, sn.as_slice(), addr.as_slice()].concat();
-        // include in fixed-length container
-        let mut key = [0u8; MAX_ADDRESS_LEN];
-        key[..sp.len()].copy_from_slice(sp.as_slice());
         Ok(Self {
-            inner: Address::new_delegated(*DEFAULT_IPC_ACTOR_NAMESPACE_ID, &key)?,
+            subnet_id: sn.clone(),
+            raw_address: *addr,
```

Comment:
Now that this cannot fail, maybe simply return `Self`?

Reply:
Same holds for other methods.

Reply:
I think Result is better, it's more general, maybe in the future, the implementation will change, but interface can still stay. If one day we have to use `Result` again, then we would end up in world of pain...

Reply:
If so then one would return `Result` almost everywhere...

Reply:
I think that `SubnetID` and `Address` should rather not be able to represent an invalid value, and `IPCAddress` should be valid for any combination of those.

Reply:
If we are sure this is the implementation to go, I think we can remove `Result`.

Reply:
We can return `Self` here, no `Result` needed (for now).